lib/kino/js.ex

```diff
@@ -0,0 +1,362 @@
+  to specify where the corresponding directory is located:
+
+      use Kino.JS, assets_path: Path.expand("assets/html", __DIR__)
+
+  ### URLs
+
+  When using multiple asset files, make sure to use relative URLs.
+  For example, when adding an image to the page, instead of:
+
+      <img src="/images/cat.jpeg" />
+
+  Do:
+
+      <img src="./images/cat.jpeg" />
+
+  This will correctly point to the `images/cat.jpeg` file in your
+  assets.
+
+  ## JavaScript API
+
+  In the example we briefly introduced the `ctx` (context) object
+  that is made available in the `init(ctx, data)` function. This
+  object encapsulates all of the Livebook-specific API that we can
+  call on the JavaScript side.
+
+  ### Properties
+
+    * `ctx.root` - the root element controlled by the widget
+
+  ### Functions
+
+    * `ctx.importCSS(url)` - loads CSS from the given URL into the
+      page. Returns a `Promise` that resolves once the CSS is loaded
+
+    * `ctx.handleEvent(event, callback)` - registers an event
+      handler. Once `event` is broadcasted, `callback` is executed
+      with the event payload. This applies to `Kino.JS.Live` widgets
+
+    * `ctx.pushEvent(event, payload)` - sends an event to the widget
+      server, where it is handled with `c:Kino.JS.Live.handle_event/3`.
+      This applies to `Kino.JS.Live` widgets
+
+  ## CDN
+
+  It is possible to use a regular JavaScript bundler for generating
+  the assets, however in many cases a simpler and preferred approach
+  is to import the necessary dependencies directly from a CDN.
+
+  To give a concrete example, here's how we could use the `mermaid`
+  JavaScript package for rendering diagrams:
+
+      defmodule Kino.Mermaid do
+        use Kino.JS
+
+        def new(graph) do
+          Kino.JS.new(__MODULE__, graph)
+        end
+
+        asset "main.js" do
+          """
+          import "https://cdn.jsdelivr.net/npm/mermaid@8.13.3/dist/mermaid.min.js";
+
+          mermaid.initialize({ startOnLoad: false });
+
+          export function init(ctx, graph) {
+            mermaid.render("graph1", graph, (svgSource, bindListeners) => {
+              ctx.root.innerHTML = svgSource;
+              bindListeners && bindListeners(ctx.root);
+            });
+          }
+          """
+        end
+      end
+
+  And we would use it like so:
+
+      Kino.Mermaid.new("""
+      graph TD;
+        A-->B;
+        A-->C;
+        B-->D;
+        C-->D;
+      """)
+
+  ## Live widgets
+
+  So far we covered the API for defining static widgets, where the
+  JavaScript side only receives the initial data and there is no
+  further interaction with the Elixir side. To introduce such
+  interaction, see `Kino.JS.Live` as a next step in our discussion.
+  '''
+
+  defstruct [:module, :data]
+
+  @type t :: %__MODULE__{module: module(), data: term()}
+
+  defmacro __using__(opts) do
+    quote location: :keep, bind_quoted: [opts: opts] do
```

Comment:
Ideally speaking you want to keep `quote` blocks as short as possible. In this case, I would do this:

```
quote do
  import Kino.JS, only: [asset: 2]
  @before_compile Kino.JS
  Kino.JS.__register__(assets_path)
end
```

`__register__` would do the job of calling `Module.put_attribute` to set external resources and default attributes.

Reply:
Beautiful, I didn't consider we can just pass the env :D

Reply:
We still need to define `__mix_recompile__?`, so something like this:

```elixir
    quote location: :keep, bind_quoted: [opts: opts] do
      import Kino.JS, only: [asset: 2]

      @before_compile Kino.JS

      assets_path = opts[:assets_path]
      asset_paths = Kino.JS.__paths__(assets_path)
      Kino.JS.__register__(__ENV__, assets_path, asset_paths)

      # Force recompilation if new assets are added
      def __mix_recompile__? do
        current_paths = Kino.JS.__paths__(unquote(assets_path))
        :erlang.md5(current_paths) != unquote(:erlang.md5(asset_paths))
      end
    end
```

Any way to simplify this further?

Reply:
You can also define it in `__before_compile__` but your approach is also great!